src/app/core/public/api/impl/Component.ts

```diff
@@ -48,14 +49,19 @@ export class ComponentImpl extends BaseObjectImpl implements Component {
     }
 
     public set angle(val: number) {
-        this.circuit.setPropFor<Schema.Component, "angle">(this.id, "angle", val);
+        this.internal.setPropFor<Schema.Component, "angle">(this.id, "angle", val);
     }
     public get angle(): number {
         return (this.getObj().props.angle ?? 0);
     }
 
-    public get ports(): Port[] {
-        return [...this.circuit.getPortsForComponent(this.id)]
-            .map((id) => new PortImpl(this.state, id));
+    public get ports(): Record<string, Port[]> {
+        return FromConcatenatedEntries(
+            [...this.internal.getPortsForComponent(this.id)]
+            .map((id) => this.circuit.constructPort(id))
+            .map((p) => [p.group, p])
+        );
     }
+
+    public abstract firstAvailable(group: string): Port | undefined;
```

Comment:
what's the reason for being abstract?  If ports have an `index` then this shouldn't depend on analog/digital, right?

Reply:
Because digital ports could be unavailable if they are input ports and already have a connection

Reply:
But that's still a problem with any kind of port.  Otherwise what's the meaning of `firstAvailable`?

Reply:
`firstAvailable` would be the first "available to be connected to" port in the given group -- we don't currently have like a `isPortAvailable` function anywhere so I guess I thought I'd do it here, where should this behavior be described then?

Reply:
Maybe I'm forgetting a use-case, but should this be on `DigitalComponent` instead?  My understanding is that for analog it would always return the first port anyway.  Is this right?

Reply:
Mmm that's a good point actually yeah, it being unique to DigitalComponent makes sense